src/filebrowser/transfer-mgr.cpp

```diff
@@ -123,19 +379,831 @@ void TransferManager::cancelAllDownloadTasks()
+        if (parent_task_id) {
+            deleteUploadTask(parent_task_id, "FolderTasks");
+        } else {
+            parent_task_id = getUploadTaskId(repo_id, path, QString(), TASK_STARTED, "FolderTasks");
+            if (parent_task_id) {
+                deleteUploadTask(parent_task_id, "FolderTasks");
+
+                const FileTaskRecord *pending_file_task = getPendingFileTask();
+                if (isValidTask(pending_file_task)) {
+                    checkUploadName(pending_file_task);
+                }
+            }
+        }
+    }
+}
+
+QList<const FileNetworkTask*> TransferManager::getTransferringTasks(
+    const QString& repo_id, const QString& parent_dir)
+{
+    QList<const FileNetworkTask*> tasks;
+
+    if (isTaskForGivenParentDir(current_download_.data(), repo_id, parent_dir)) {
         tasks.append(current_download_.data());
     }
+    if (isTaskForGivenParentDir(current_upload_.data(), repo_id, parent_dir)) {
+        tasks.append(current_upload_.data());
+    }
+
     foreach (const QSharedPointer<FileDownloadTask>& task, pending_downloads_) {
-        if (isDownloadForGivenParentDir(task, repo_id, parent_dir)) {
+        if (isTaskForGivenParentDir(task.data(), repo_id, parent_dir)) {
             tasks.append(task.data());
         }
     }
 
     return tasks;
 }
+
+QList<const FileTaskRecord*> TransferManager::getPendingUploadFiles(
+    const QString& repo_id, const QString& parent_dir)
+{
+    QList<const FileTaskRecord*> ret;
+
+    QMap<int, FileTaskRecord>::const_iterator ite = file_tasks_.constBegin();
+    while (ite != file_tasks_.constEnd()) {
+        const FileTaskRecord* task = &ite.value();
+        if (task->status == kPendingStatus &&
+            task->repo_id == repo_id &&
+            ::getParentPath(task->path) == parent_dir) {
+            ret.push_back(task);
+        }
+        ++ite;
+    }
+
+    return ret;
+}
+
+QList<const FolderTaskRecord*> TransferManager::getUploadFolderTasks(
+    const QString& repo_id, const QString& parent_dir)
+{
+    QList<const FolderTaskRecord*> ret;
+
+    QMap<int, FolderTaskRecord>::const_iterator ite = folder_tasks_.constBegin();
+    while (ite != folder_tasks_.constEnd()) {
+        const FolderTaskRecord* task = &ite.value();
+        if (task->repo_id == repo_id &&
+            ::getParentPath(task->path) == parent_dir) {
+            ret.push_back(task);
+        }
+        ++ite;
+    }
+
+    return ret;
+}
+
+bool TransferManager::isTransferring(const QString& repo_id,
+                                     const QString& path)
+{
+    if (getDownloadTask(repo_id, path) ||
+        isCurrentUploadTask(repo_id, path) ||
+        getUploadTaskId(repo_id, path)) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+void TransferManager::cancelTransfer(const QString& repo_id,
+                                     const QString& path)
+{
+    cancelDownload(repo_id, path);
+    cancelUpload(repo_id, path);
```

Comment:
This is not good code style. Be clear about the logic: either cancel a download, or cancel a upload.